### ISSUE TITLE
the evaluation result of magicoder is not aligned with the result on paper

### ISSUE THREAD

**Issue by maowayne123** [ise-uiuc/magicoder#16]
Hi, thanks for your great work.

I test the performance of magicoder, however the performance is not align with the result on the paper(68.9 vs. 76.8). I guess it is because I used different hyperparameter for inference, eg,  --top_p 1, --temperature 1 and so on. I will be grateful if author can provide the specific hyperparameter for inference. Thank u. i list the script i used below:

I try to evaluate the magicoder on the script you provided in the 'experiment' folder by the following commond:
```
python experiments/text2code.py \
    --model_key deepseek-ai/deepseek-coder-6.7b-base \
    --dataset humaneval \
    --save_path output_dir/mc_6_7_ds \
    --n_batches 1 \
    --n_problems_per_batch 1 \
    --n_samples_per_problem 1 \
    --model_name_or_path ~/weight/magicoders-s-ds-6.7/ \
    --top_p 1 \
    --max_new_tokens 4096 \
    --temperature 1
```
then I use the the the following commond for evalplus:
```
docker run -v $(pwd):/app ganler/evalplus:latest --dataset humaneval --samples output_dir/mc_6_7_ds.jsonl
```

Finally, i got the result:
```
Base
{'pass@1': 0.6890243902439024}
Base + Extra
{'pass@1': 0.6158536585365854}
```

**Comment by UniverseFly** [comment]
Hi @maowayne123, in our paper we use greedy decoding for all HumanEval experiments, which means the temperature is 0. The exact hyper parameters are

```
    --temperature 0.0
    --top_p 1.0
    --max_new_tokens 512
    --n_problems_per_batch 16
    --n_samples_per_problem 1
    --n_batches 1
 ```
 
The choice of `n_problems_per_batch` will only slightly affect the results due to floating points round-off (related discussion: https://discuss.pytorch.org/t/results-of-forward-pass-are-different-with-different-batch-size/162277).

**Comment by maowayne123** [comment]
> Hi @maowayne123, in our paper we use greedy decoding for all HumanEval experiments, which means the temperature is 0. The exact hyper parameters are
> 
> ```
>     --temperature 0.0
>     --top_p 1.0
>     --max_new_tokens 512
>     --n_problems_per_batch 16
>     --n_samples_per_problem 1
>     --n_batches 1
> ```
> 
> The choice of `n_problems_per_batch` will only slightly affect the results due to floating points round-off (related discussion: https://discuss.pytorch.org/t/results-of-forward-pass-are-different-with-different-batch-size/162277).

Hi, thanks for your reply. I try your hyperparameters. This time I got much better result, but the result still not align with the paper result (75.6% vs. 76.8%). do u have any idea for it? the script I used this time is:
```
python experiments/text2code.py \
    --model_key deepseek-ai/deepseek-coder-6.7b-base \
    --dataset humaneval \
    --save_path output_dir/mc_6_7_ds_flash_attn.jsonl \
    --n_batches 1 \
    --n_problems_per_batch 2 \
    --n_samples_per_problem 1 \
    --model_name_or_path ~/weight/magicoders-s-ds-6.7/ \
    --top_p 1.0 \
    --max_new_tokens 512 \
    --temperature 0.0
```
BTW, currently the BFloat16 is enable. This may cause the precision loss. Moreover, flash attention can result in better result (75.6% vs. 75.0%). Thank you!

**Comment by maowayne123** [comment]
I found a more embarrassing things. The deepseeker can reach  81.7% acc. there must be something wrong.

The script I used is:
```
python experiments/text2code.py \
    --model_key deepseek-ai/deepseek-coder-6.7b-base \
    --dataset humaneval \
    --save_path output_dir/ds_6_7_ds.jsonl \
    --n_batches 1 \
    --n_problems_per_batch 2 \
    --n_samples_per_problem 1 \
    --model_name_or_path ~/weight/deepseek-coder-6.7b/ \
    --top_p 1.0 \
    --max_new_tokens 512 \
    --temperature 0.0
```
the test script i used is:
```
docker run -v $(pwd):/app ganler/evalplus:latest --dataset humaneval --samples output_dir/ds_6_7_ds.jsonl
```
the result i got is:
```
Base
{'pass@1': 0.8170731707317073}
Base + Extra
{'pass@1': 0.7621951219512195}
```

**Comment by UniverseFly** [comment]
@maowayne123 Thanks for reporting this interesting finding. I suppose you were running the DeepSeek-Coder Instruct version? We didn't run it using the script but instead report the [EvalPlus Leaderboard](https://evalplus.github.io/leaderboard.html) results. If your finding is true, the prompt could be the factor. Also, since DeepSeek-Coder does not release its instruction data, we cannot analyze if there are contamination issues.

**Comment by UniverseFly** [comment]
And the results discrepancy can be due to your choice of `n_problems_per_batch`, which influences cuBLAS's optimization strategy. If you experiment with different values, you may see different results. We choose `16` for a faster evaluation.

**Comment by answers111** [comment]
Hello, thank you for such an impressing work.
I'm trying to reproduce the result score reported in your paper. Using the script you provided in the 'experiment' folder, I got

```
Base
{'pass@1': 0.6402439024390244}
Base + Extra
{'pass@1': 0.573170731707317}
``` 
for Magicoder-DS-6.7B and

```
Base
{'pass@1': 0.7439024390243902}
Base + Extra
{'pass@1': 0.6890243902439024}
``` 
for Magicoder-S-DS-6.7B which is not aligned with the result on paper.

 the commond i used is
```
python experiments/text2code.py \
    --model_key deepseek-ai/deepseek-coder-6.7b-base \
    --dataset humaneval \
    --save_path ds_6_7_ds.jsonl \
    --n_batches 1 \
    --n_problems_per_batch 16 \
    --n_samples_per_problem 1 \
    --model_name_or_path ise-uiuc/Magicoder-DS-6.7B \
    --top_p 1.0 \
    --max_new_tokens 512 \
    --temperature 0.0 
```
I'm confuse about it.  Would be grateful if you can provide some extra information. Thank u.

**Comment by UniverseFly** [comment]
Hi @maowayne123 and @answers111, I deeply appreciate your efforts in running Magicoder evaluations and raising relevant issues. I wrote the documentation on how to **100%** reproduce the paper results on HumanEval(+) and MBPP(+) [here](https://github.com/ise-uiuc/magicoder/blob/7fd470461ccb1adcdf959192c742e8361ab6d452/experiments/README.md).
I am also attaching the generated samples here: [magicoder-evalplus-results.tar.gz](https://github.com/ise-uiuc/magicoder/files/13746794/magicoder-evalplus-results.tar.gz).

Upon investigation, the reasons for the discrepancy are two folds: **package versions** and **batch size**. During the experiments, we just chose the batch size that can best utilize available GPUs (including 16, 24, and 28). We actually didn't pay too much attention to the randomness incurred by its choice. Choosing a different value can slightly **improve/worsen** the results. I also drafted a separate Limitations section in the README based on your helpful comments.

Anyways, great thanks to your participations in the discussion. I will close this issue for now but feel free to reopen it in case of any questions. We will further improve Magicoder in the near future and provide more comprehensive evaluation!
